### ISSUE TITLE
Microstutters on Windows 10.0.18363

### ISSUE THREAD

**Issue by harshavb** [ppy/osu#12055]
**Describe the bug:** 
I get small microstutters when playing osu!lazer. Although the microstutters are small, they certainly aren't unnoticable, and I don't face the same issue in stable.

**Screenshots or videos showing encountered issue:** 
A video is available [here](https://www.youtube.com/watch?v=gJTP1999Gg0).

**osu!lazer version:** 

**Logs:**
[performance.log](https://github.com/ppy/osu/files/6161081/performance.log)
[network.log](https://github.com/ppy/osu/files/6161082/network.log)
[runtime.log](https://github.com/ppy/osu/files/6161083/runtime.log)
[updater.log](https://github.com/ppy/osu/files/6161084/updater.log)

**Comment by smoogipoo** [comment]
Hard to say. The few spikes throughout are GC spikes, which are related to the Linux issue, and ultimately finalizers in general. The final spike you point out is just a general draw spike and is probably more related to overall GPU load.

**Comment by harshavb** [comment]
In this case should we reserve this issue for the draw spikes? I'm getting both gc and draw spikes, but it's quite likely that any fixes for the gc spikes in the other issue will apply for mine as well.

**Comment by smoogipoo** [comment]
Yes the GC fixes will apply to you too. We're already tracking that one in https://github.com/ppy/osu-framework/issues/4229 and the dotnet/runtime issue linked there.

So this one can be repurposed for the draw spike, but I also don't know whether there is a fix that can be done here, it needs more investigation.

**Comment by sr229** [comment]
Going to raise this for Windows Dev Build 21032 as well. It wasn't just a microsttuter but at some point in the game there would be a 2-second "hangup", probably GC doing its thing. Will check if the o!f PR patch would resolve this.

**Comment by peppy** [comment]
@sr229 comments like yours are basically spam at this point. unless you are going to provide logs, information, or add something, please refrain from posting.

**Comment by sr229** [comment]
> @sr229 comments like yours are basically spam at this point. unless you are going to provide logs, information, or add something, please refrain from posting.

Oh, I thought I uploaded it for some reason, I attached the logs here now.


[performance-update.log](https://github.com/ppy/osu/files/6172836/performance-update.log)
[runtime.log](https://github.com/ppy/osu/files/6172838/runtime.log)
[network.log](https://github.com/ppy/osu/files/6172839/network.log)
[performance.log](https://github.com/ppy/osu/files/6172840/performance.log)
[performance-audio.log](https://github.com/ppy/osu/files/6172841/performance-audio.log)
[performance-draw.log](https://github.com/ppy/osu/files/6172842/performance-draw.log)

Some of the logs date back to 2021.302 but may be relevant.

**Comment by Terratype** [comment]
I'm having this issue to the point of unplayability. The entire time the game is running for me there are stutters every 1-2 seconds. I'm not sure if this is related but I had a similar issue in stable that was fixed by turning off windows game mode.

[input.log](https://github.com/ppy/osu/files/9856584/input.log)
[network.log](https://github.com/ppy/osu/files/9856585/network.log)
[performance.log](https://github.com/ppy/osu/files/9856587/performance.log)
[runtime.log](https://github.com/ppy/osu/files/9856588/runtime.log)
[updater.log](https://github.com/ppy/osu/files/9856589/updater.log)
[database.log](https://github.com/ppy/osu/files/9856590/database.log)

**Comment by peppy** [comment]
It's going to be something else on your system.

https://osu.ppy.sh/help/wiki/Performance_Troubleshooting may have some tips you can follow. Note that turning *off* game mode should never be a solution.

**Comment by AlfredTheSalmon** [comment]
I seem to be having a similar issue, on a very periodic interval. They are defenitely noticeable during gameplay, and I always had them throughout the last few months. A video is available [here](https://streamable.com/3e2pd4). The stutters appear to be synchronous to all threads, so I suspect the issue is still GC.

[network.log](https://github.com/ppy/osu/files/13732857/network.log)
[performance.log](https://github.com/ppy/osu/files/13732858/performance.log)
[runtime.log](https://github.com/ppy/osu/files/13732859/runtime.log)
[updater.log](https://github.com/ppy/osu/files/13732860/updater.log)
[database.log](https://github.com/ppy/osu/files/13732861/database.log)
[input.log](https://github.com/ppy/osu/files/13732862/input.log)

**Comment by smoogipoo** [comment]
Does using the OpenGL renderer improve things for you?

**Comment by AlfredTheSalmon** [comment]
It helps, the draw fps is much more stable, but it still happens. In this [video](https://streamable.com/ov1p97) (the gameplay itself starts at around :40 and the fps tanks because OBS, usually stays between 650 and 700) I also reduced the limiter to 4x and changed the skin to argon to eliminate possible variables. 
Im not discarding the fact that it may also be a problem of my pc configuration, i have a 5800X3D + 6750XT, running on:
- windows 11 version 10.0.22621 Build 22621
- Latest AMD Adrenaline drivers (AMD Windows Driver Version 31.0.23013.1023 and Driver Version 23.30.13.01-231128a-398226C-AMD-Software-Adrenalin-Edition)
- Direct3D® Driver Version 9.14.10.01526
- OpenGL® Driver Version 23.11.230906_ab0988b

**Comment by dotremilie** [comment]
Hi, I have similar problem when running on Direct3D. Generally there is a visible spike every ~2 seconds. ([https://streamable.com/gs8dgo](video)) Additionally when you open the mod menu it halves the framerate when you move. The issue seem to not be that harsh on OpenGL, general performance is better and I have stable 1000fps while recording. On D3D it's around 600, 8x limiter. PC specifications should not be a problem in terms of performance because I have 7800X3D/7900XT. I made sure I have latest drivers, Windows build etc. 

Logs:
[1704325081.performance-draw.log](https://github.com/ppy/osu/files/13824551/1704325081.performance-draw.log)
[1704325081.performance-input.log](https://github.com/ppy/osu/files/13824552/1704325081.performance-input.log)
[1704325081.performance-update.log](https://github.com/ppy/osu/files/13824553/1704325081.performance-update.log)
[1704325081.runtime.log](https://github.com/ppy/osu/files/13824554/1704325081.runtime.log)
[1704325081.updater.log](https://github.com/ppy/osu/files/13824555/1704325081.updater.log)
[1704325081.database.log](https://github.com/ppy/osu/files/13824556/1704325081.database.log)
[1704325081.input.log](https://github.com/ppy/osu/files/13824557/1704325081.input.log)
[1704325081.network.log](https://github.com/ppy/osu/files/13824558/1704325081.network.log)
[1704325081.performance.log](https://github.com/ppy/osu/files/13824559/1704325081.performance.log)
[1704325081.performance-audio.log](https://github.com/ppy/osu/files/13824560/1704325081.performance-audio.log)

**Comment by SupDos** [comment]
Wanted to also hop in to say I have a similar issue, and it seems all 3 of us have full AMD CPU/GPU (5800x3D, 6700XT here)
To add to this, running the game on Linux on the same machine completely fixes the stutters, and there are no spikes visible on the draw thread
I also tried fully reinstalling Windows from scratch, and I still get the same spikes on a clean install

[compressed-logs.zip](https://github.com/ppy/osu/files/13830658/compressed-logs.zip)

**Comment by peppy** [comment]
Can you reproduce the stutters during autoplay? Do they stop when pausing?

**Comment by SupDos** [comment]
Yes, forgot to add that the stutters are only during gameplay/autoplay, pausing causes them to stop (even when moving mouse around and pressing keys while paused) and there are none of these stutters anywhere else in the game

**Comment by peppy** [comment]
Thanks, I can reproduce this (albeit less severe) and am investigating.

Can you also test with the classic skin with no modifications?

**Comment by SupDos** [comment]
With classic skin I feel like I notice it less, but I’m not sure if that’s just because I don’t notice it with the amount of extra “fluff” the skin has

In any case, I can definitely still see green spikes on the draw thread with classic, but they seem to mostly be half or 3/4 spikes, and I usually only notice the stutter when the green spike reaches the top of the graph (if that makes any sense)

Will continue testing along with trying the other skins as well and get back with more info if I find anything else

**Comment by peppy** [comment]
Why are you running on OpenGL renderer? This is the least supported renderer of all, and is a likely cause. Please set your renderer to direct x.

**Comment by SupDos** [comment]
OpenGL is the only one that allows true exclusive fullscreen (DirectX forces fullscreen optimisations on, so borderless, no matter what)

Either way this happens on all 3 renders, I still get the spikes no matter what renderer it is, and @AlfredTheSalmon above who was asked to try OpenGL also mentioned that they still got the stutters there

(Unrelated to the current issue but that might need rewording, I always thought that since it said “Legacy” ie old, the non-legacy version should be used instead)

**Comment by dotremilie** [comment]
For me the stutters don't happen (or are not noticable) on OpenGL (Legacy as well), only on DirectX. Framerate is basically halved compared to OpenGL. Just now I alt-tabed back to lazer (DirectX) to see it's 50-80fps in menu, and 200-300 in gameplay. Here are all the logs. 

[compressed-logs.zip](https://github.com/ppy/osu/files/13833225/compressed-logs.zip)

I played all the time on DirectX, just switched to OpenGL for testing and it performs way better, I as well thought Legacy = outdated/don't use by the way.
